File: clients/go-sdk/scheduledjobs/runner.go
// Package scheduledjobs implements the consumer-side dispatcher for
// platform-fired scheduled-job webhooks.
//
// Mount Runner.Process on whatever HTTP framework you use at the URL
// you set as target_url on the job definition. Each platform → SDK
// firing arrives as a JSON envelope; Process validates it, kicks off
// the handler in the background, and returns a Result immediately so
// you can map to HTTP status without blocking on the handler.
//
// Concurrency control is via an injected lock.Provider. For
// concurrent: false jobs (the platform doesn't enforce this — see
// CLAUDE.md), the lock key defaults to "scheduled-job:{jobCode}";
// contended fires return immediately without invoking the handler,
// and if the job tracks completion, a Failure with
// reason "lock-held" is reported back.
//
// # Example
//
//	r := scheduledjobs.NewBuilder(apiClient, lockProvider).
//	    Handler("daily-rollup", func(ctx context.Context, hctx *scheduledjobs.HandlerContext) (json.RawMessage, error) {
//	        hctx.Log(ctx, "starting", nil)
//	        // … work …
//	        return json.RawMessage(`{"processed":42}`), nil
//	    }).
//	    Build()
//
//	http.HandleFunc("/webhooks/scheduled-job", func(w http.ResponseWriter, req *http.Request) {
//	    var raw json.RawMessage
//	    json.NewDecoder(req.Body).Decode(&raw)
//	    switch res := r.Process(raw); res.Kind {
//	        case scheduledjobs.ResultAccepted:  w.WriteHeader(http.StatusAccepted)
//	        case scheduledjobs.ResultNotFound:  http.Error(w, res.Message, http.StatusNotFound)
//	        case scheduledjobs.ResultBadRequest: http.Error(w, res.Message, http.StatusBadRequest)
//	    }
//	})
package scheduledjobs

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/flowcatalyst/flowcatalyst/clients/go-sdk/client"
	"github.com/flowcatalyst/flowcatalyst/clients/go-sdk/lock"
)

// DefaultLockTTL is how long a job's serialisation lock is held when
// not overridden via Builder.LockTTL. Should comfortably exceed your
// expected critical-section duration; the lock self-releases on
// handler completion.
const DefaultLockTTL = 10 * time.Minute

// MaxResultBytes caps the size of completion result JSON forwarded to
// the platform; oversized payloads are replaced with a preview.
const MaxResultBytes = 10_000

// TriggerKind identifies how a firing was started.
type TriggerKind string

const (
	TriggerCron   TriggerKind = "CRON"
	TriggerManual TriggerKind = "MANUAL"
)

// Envelope is the JSON body POSTed by the platform's scheduler
// dispatcher. Mirrors ScheduledJobEnvelope in the TS / Rust SDKs.
type Envelope struct {
	JobID            string          `json:"jobId"`
	JobCode          string          `json:"jobCode"`
	InstanceID       string          `json:"instanceId"`
	ScheduledFor     string          `json:"scheduledFor,omitempty"`
	FiredAt          string          `json:"firedAt"`
	TriggerKind      TriggerKind     `json:"triggerKind"`
	CorrelationID    string          `json:"correlationId,omitempty"`
	Payload          json.RawMessage `json:"payload,omitempty"`
	TracksCompletion bool            `json:"tracksCompletion"`
	TimeoutSeconds   *uint32         `json:"timeoutSeconds,omitempty"`
}

// HandlerFunc is invoked once per accepted firing. Return a non-nil
// JSON value on success (stored as completion_result on the
// instance); return a non-nil error on failure (the error message is
// stored). Nil JSON + nil error → empty success.
type HandlerFunc func(ctx context.Context, hctx *HandlerContext) (json.RawMessage, error)

// HandlerContext is passed to user handlers. Wraps the envelope and a
// best-effort Log callback that posts a log line back to the
// platform.
type HandlerContext struct {
	Envelope Envelope

	api        *client.FlowCatalystClient
	instanceID string
	jobCode    string
	onError    func(err error, env *Envelope)
}

// Log appends a structured log entry to this instance. Best-effort —
// errors are forwarded to the runner's OnError hook (if configured)
// and then swallowed; logging failures never fail the handler.
func (h *HandlerContext) Log(ctx context.Context, message string, opts *LogOptions) {
	req := &client.InstanceLogRequest{Message: message}
	if opts != nil {
		req.Level = client.LogLevel(opts.Level)
		req.Metadata = opts.Metadata
	}
	if _, err := h.api.ScheduledJobs().LogForInstance(ctx, h.instanceID, req); err != nil {
		if h.onError != nil {
			h.onError(&RunnerError{Kind: ErrCallbackFailed, Cause: err}, &h.Envelope)
		}
	}
}

// LogOptions configures HandlerContext.Log. Level defaults to INFO
// when nil or empty.
type LogOptions struct {
	Level    string
	Metadata json.RawMessage
}

// ResultKind classifies the outcome of Runner.Process.
type ResultKind int

const (
	// ResultAccepted — envelope recognised; handler dispatched in
	// the background. Respond with HTTP 202.
	ResultAccepted ResultKind = iota
	// ResultBadRequest — envelope malformed. Respond with HTTP 400.
	ResultBadRequest
	// ResultNotFound — envelope well-formed but no handler is
	// registered for the job code. Respond with HTTP 404.
	ResultNotFound
)

// Result is the return value of Runner.Process. Map to your
// framework's HTTP response type.
type Result struct {
	Kind    ResultKind
	Message string
}

// ErrorKind classifies runner-internal failures surfaced via the
// OnError hook.
type ErrorKind int

const (
	// ErrCallbackFailed — log or complete API call failed.
	ErrCallbackFailed ErrorKind = iota
	// ErrLockFailed — lock acquire or release failed.
	ErrLockFailed
	// ErrHandlerPanicked — the user handler returned an error or
	// panicked (recovered).
	ErrHandlerPanicked
)

// RunnerError wraps a runner-internal failure for the OnError hook.
type RunnerError struct {
	Kind  ErrorKind
	Cause error
}

func (e *RunnerError) Error() string {
	prefix := "scheduledjobs: "
	switch e.Kind {
	case ErrCallbackFailed:
		prefix += "callback failed"
	case ErrLockFailed:
		prefix += "lock failed"
	case ErrHandlerPanicked:
		prefix += "handler panicked"
	}
	if e.Cause != nil {
		return prefix + ": " + e.Cause.Error()
	}
	return prefix
}

func (e *RunnerError) Unwrap() error { return e.Cause }

// OnErrorHook is fired on every uncaught handler / callback / lock
// error. The hook runs synchronously inside the background goroutine;
// if you need async reporting, dispatch from here to your own
// channel.
type OnErrorHook func(err error, env *Envelope)

// LockKeyFunc derives the lock key from an envelope. The default is
// "scheduled-job:{jobCode}".
type LockKeyFunc func(env *Envelope) string

// Runner is the registry-of-handlers + envelope dispatcher.
type Runner struct {
	api          *client.FlowCatalystClient
	lockProvider lock.Provider
	handlers     map[string]HandlerFunc
	opts         runnerOpts

	wg sync.WaitGroup // tracks background goroutines for Shutdown
}

type runnerOpts struct {
	enforceLock bool
	lockTTL     time.Duration
	lockKey     LockKeyFunc
	onError     OnErrorHook
	background  func(fn func()) // test seam: defaults to go fn()
}

// Builder configures a Runner. Start one with NewBuilder.
type Builder struct {
	api          *client.FlowCatalystClient
	lockProvider lock.Provider
	handlers     map[string]HandlerFunc
	opts         runnerOpts
}

// NewBuilder starts a builder. Pass the API client used for
// log/complete callbacks and a lock.Provider (use lock.NewNoOp() if
// you don't need concurrency control).
func NewBuilder(api *client.FlowCatalystClient, lockProvider lock.Provider) *Builder {
	return &Builder{
		api:          api,
		lockProvider: lockProvider,
		handlers:     map[string]HandlerFunc{},
		opts: runnerOpts{
			enforceLock: true,
			lockTTL:     DefaultLockTTL,
			lockKey:     func(e *Envelope) string { return "scheduled-job:" + e.JobCode },
		},
	}
}

// Handler registers a handler keyed by the job's code.
func (b *Builder) Handler(code string, h HandlerFunc) *Builder {
	b.handlers[code] = h
	return b
}

// EnforceLock toggles whether Process acquires a lock before
// dispatching. Default: true.
func (b *Builder) EnforceLock(enforce bool) *Builder { b.opts.enforceLock = enforce; return b }

// LockTTL overrides DefaultLockTTL.
func (b *Builder) LockTTL(ttl time.Duration) *Builder { b.opts.lockTTL = ttl; return b }

// LockKey overrides the default key function.
func (b *Builder) LockKey(f LockKeyFunc) *Builder { b.opts.lockKey = f; return b }

// OnError sets the failure hook.
func (b *Builder) OnError(h OnErrorHook) *Builder { b.opts.onError = h; return b }

// Build finalises the Runner.
func (b *Builder) Build() *Runner {
	return &Runner{
		api:          b.api,
		lockProvider: b.lockProvider,
		handlers:     b.handlers,
		opts:         b.opts,
	}
}

// RegisteredCodes returns the set of registered handler codes
// (diagnostics only).
func (r *Runner) RegisteredCodes() []string {
	out := make([]string, 0, len(r.handlers))
	for k := range r.handlers {
		out = append(out, k)
	}
	return out
}

// Process is the inbound dispatch entry point. Validates raw as an
// Envelope, kicks off the handler in the background, and returns a
// Result immediately. The actual handler + lock + completion
// callback run on a goroutine.
func (r *Runner) Process(raw json.RawMessage) Result {
	var env Envelope
	if err := json.Unmarshal(raw, &env); err != nil {
		return Result{Kind: ResultBadRequest, Message: err.Error()}
	}
	if env.TriggerKind != TriggerCron && env.TriggerKind != TriggerManual {
		return Result{Kind: ResultBadRequest, Message: "invalid triggerKind"}
	}
	handler, ok := r.handlers[env.JobCode]
	if !ok {
		return Result{Kind: ResultNotFound, Message: fmt.Sprintf("no handler registered for code %q", env.JobCode)}
	}

	r.wg.Add(1)
	dispatch := r.opts.background
	if dispatch == nil {
		dispatch = func(fn func()) { go fn() }
	}
	dispatch(func() {
		defer r.wg.Done()
		r.runInBackground(env, handler)
	})
	return Result{Kind: ResultAccepted}
}

// Wait blocks until all in-flight background dispatches finish.
// Intended for tests + graceful shutdown.
func (r *Runner) Wait() { r.wg.Wait() }

func (r *Runner) runInBackground(env Envelope, handler HandlerFunc) {
	ctx := context.Background()
	if env.TimeoutSeconds != nil && *env.TimeoutSeconds > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, time.Duration(*env.TimeoutSeconds)*time.Second)
		defer cancel()
	}

	var held lock.Handle
	if r.opts.enforceLock && r.lockProvider != nil {
		key := r.opts.lockKey(&env)
		h, err := r.lockProvider.Acquire(ctx, key, r.opts.lockTTL)
		if err != nil {
			r.fireOnError(&RunnerError{Kind: ErrLockFailed, Cause: err}, &env)
			return
		}
		if h == nil {
			// Lock contention — skip. If the job tracks completion,
			// mark Failure with reason "lock-held" so the operator
			// can see the firing was skipped, not silently lost.
			if env.TracksCompletion {
				r.reportSkipped(ctx, &env)
			}
			return
		}
		held = h
	}
	defer func() {
		if held != nil {
			if err := held.Release(ctx); err != nil {
				r.fireOnError(&RunnerError{Kind: ErrLockFailed, Cause: err}, &env)
			}
		}
	}()

	hctx := &HandlerContext{
		Envelope:   env,
		api:        r.api,
		instanceID: env.InstanceID,
		jobCode:    env.JobCode,
		onError:    r.opts.onError,
	}

	value, handlerErr := r.invokeHandler(ctx, handler, hctx, &env)

	if env.TracksCompletion {
		req := &client.InstanceCompleteRequest{}
		if handlerErr == nil {
			req.Status = client.CompletionStatusSuccess
			req.Result = sanitiseResult(value)
		} else {
			req.Status = client.CompletionStatusFailure
			req.Result = json.RawMessage(fmt.Sprintf(`{"error":%q}`, handlerErr.Error()))
		}
		if _, err := r.api.ScheduledJobs().CompleteInstance(ctx, env.InstanceID, req); err != nil {
			r.fireOnError(&RunnerError{Kind: ErrCallbackFailed, Cause: err}, &env)
		}
	}

	if handlerErr != nil {
		r.fireOnError(&RunnerError{Kind: ErrHandlerPanicked, Cause: handlerErr}, &env)
	}
}

func (r *Runner) invokeHandler(
	ctx context.Context,
	handler HandlerFunc,
	hctx *HandlerContext,
	env *Envelope,
) (value json.RawMessage, err error) {
	defer func() {
		if rec := recover(); rec != nil {
			switch v := rec.(type) {
			case error:
				err = fmt.Errorf("handler panic: %w", v)
			default:
				err = fmt.Errorf("handler panic: %v", v)
			}
			_ = env // captured for parity with hook callers; nothing else to do here
		}
	}()
	return handler(ctx, hctx)
}

func (r *Runner) reportSkipped(ctx context.Context, env *Envelope) {
	req := &client.InstanceCompleteRequest{
		Status: client.CompletionStatusFailure,
		Result: json.RawMessage(`{"skipped":true,"reason":"lock-held"}`),
	}
	if _, err := r.api.ScheduledJobs().CompleteInstance(ctx, env.InstanceID, req); err != nil {
		r.fireOnError(&RunnerError{Kind: ErrCallbackFailed, Cause: err}, env)
	}
}

func (r *Runner) fireOnError(err *RunnerError, env *Envelope) {
	if r.opts.onError != nil {
		r.opts.onError(err, env)
	}
}

func sanitiseResult(v json.RawMessage) json.RawMessage {
	if len(v) <= MaxResultBytes {
		return v
	}
	preview := string(v[:MaxResultBytes])
	return json.RawMessage(fmt.Sprintf(`{"truncated":true,"preview":%q}`, preview))
}

File: crates/fc-sdk/src/scheduled_jobs/runner.rs
//! `ScheduledJobRunner` — consumer-side dispatch for platform-fired
//! scheduled-job webhooks.
//!
//! Mirrors `@flowcatalyst/sdk`'s TypeScript `ScheduledJobRunner`. Mount
//! [`ScheduledJobRunner::process`] on whatever HTTP framework you use
//! (Axum, Actix, Rocket, raw hyper) at the URL you set as `target_url`
//! on the job definition.
//!
//! Two outputs from [`process`](ScheduledJobRunner::process):
//!   1. The HTTP response shape (always 202 Accepted on a recognised envelope).
//!   2. A background future, started but not awaited — the actual handler
//!      execution + completion callback. The platform expects 202 within the
//!      dispatcher's `http_timeout` (default 10s); your handler should run
//!      asynchronously, not block the HTTP response.
//!
//! The runner enforces concurrency via an injected [`LockProvider`]. For
//! `concurrent: false` jobs (the platform doesn't enforce this — see
//! CLAUDE.md), the lock key defaults to `scheduled-job:{jobCode}`;
//! contended fires return immediately without invoking the handler.

use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;
use std::time::Duration;

use serde::{Deserialize, Serialize};

use crate::client::scheduled_jobs::{
    CompletionStatus, InstanceCompleteRequest, InstanceLogRequest, LogLevel,
};
use crate::client::{ClientError, FlowCatalystClient};
use crate::lock::LockProvider;

const DEFAULT_LOCK_TTL: Duration = Duration::from_secs(10 * 60);
const MAX_RESULT_BYTES: usize = 10_000;

// ── Envelope ──────────────────────────────────────────────────────────────

/// Envelope POSTed by the platform's scheduler dispatcher. Mirrors
/// `ScheduledJobEnvelope` in the TypeScript SDK.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledJobEnvelope {
    pub job_id: String,
    pub job_code: String,
    pub instance_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub scheduled_for: Option<String>,
    pub fired_at: String,
    pub trigger_kind: TriggerKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<serde_json::Value>,
    pub tracks_completion: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout_seconds: Option<u32>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum TriggerKind {
    Cron,
    Manual,
}

// ── Handler context ──────────────────────────────────────────────────────

/// Context passed to user handlers. Wraps the envelope and a best-effort
/// `log` callback that posts a log line back to the platform.
#[derive(Clone)]
pub struct HandlerContext {
    pub envelope: ScheduledJobEnvelope,
    inner: Arc<HandlerContextInner>,
}

struct HandlerContextInner {
    client: FlowCatalystClient,
    instance_id: String,
    job_code: String,
    on_error: Option<OnErrorHook>,
}

impl HandlerContext {
    /// Append a structured log entry to this instance. Best-effort — errors
    /// are forwarded to the runner's `on_error` hook (if configured) and
    /// then swallowed; logging failures never fail the handler.
    pub async fn log(&self, message: impl Into<String>, opts: LogOptions) {
        let req = InstanceLogRequest {
            message: message.into(),
            level: opts.level,
            metadata: opts.metadata,
        };
        if let Err(e) = self
            .inner
            .client
            .scheduled_jobs()
            .log_for_instance(&self.inner.instance_id, &req)
            .await
        {
            if let Some(hook) = &self.inner.on_error {
                hook(
                    RunnerError::CallbackFailed(e),
                    &ScheduledJobEnvelope {
                        job_id: self.envelope.job_id.clone(),
                        job_code: self.inner.job_code.clone(),
                        instance_id: self.inner.instance_id.clone(),
                        scheduled_for: self.envelope.scheduled_for.clone(),
                        fired_at: self.envelope.fired_at.clone(),
                        trigger_kind: self.envelope.trigger_kind,
                        correlation_id: self.envelope.correlation_id.clone(),
                        payload: self.envelope.payload.clone(),
                        tracks_completion: self.envelope.tracks_completion,
                        timeout_seconds: self.envelope.timeout_seconds,
                    },
                );
            }
        }
    }
}

/// Optional fields for [`HandlerContext::log`]. Defaults: `INFO`, no metadata.
#[derive(Debug, Clone, Default)]
pub struct LogOptions {
    pub level: LogLevel,
    pub metadata: Option<serde_json::Value>,
}

// ── Runner ──────────────────────────────────────────────────────────────

/// User handler. Returns `Ok(value)` on success (the value is stored as
/// `completion_result`); `Err` on failure (the error message is stored).
pub type HandlerFuture =
    Pin<Box<dyn Future<Output = Result<serde_json::Value, HandlerError>> + Send>>;

/// Boxed handler. Closures `Fn(HandlerContext) -> HandlerFuture` are accepted;
/// see [`ScheduledJobRunner::handler`].
pub type BoxedHandler =
    Arc<dyn Fn(HandlerContext) -> HandlerFuture + Send + Sync + 'static>;

/// Error wrapper returned from handlers. Use [`HandlerError::msg`] for
/// ad-hoc string errors, or `?` to bubble any error that implements
/// `Into<HandlerError>`.
#[derive(Debug)]
pub struct HandlerError {
    pub message: String,
}

impl HandlerError {
    pub fn msg(s: impl Into<String>) -> Self {
        Self { message: s.into() }
    }
}

impl<E: std::error::Error> From<E> for HandlerError {
    fn from(e: E) -> Self {
        Self {
            message: e.to_string(),
        }
    }
}

/// Errors surfaced via the `on_error` hook (post-handler-completion failures
/// — log/complete callbacks failing, lock release failing, etc.).
#[derive(Debug)]
pub enum RunnerError {
    CallbackFailed(ClientError),
    LockReleaseFailed(String),
    HandlerPanicked(String),
}

/// Hook fired on every uncaught handler / callback / release error. The
/// hook runs synchronously inside the background task; if you need async
/// reporting, dispatch to your own channel from here.
pub type OnErrorHook = Arc<dyn Fn(RunnerError, &ScheduledJobEnvelope) + Send + Sync + 'static>;

/// Result of [`ScheduledJobRunner::process`]. The runner is HTTP-framework-
/// agnostic; map this to your framework's response type.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RunResult {
    /// Recognised envelope; handler dispatched in the background.
    /// Respond with HTTP 202 Accepted + body `{"ok": true}`.
    Accepted,
    /// Envelope was malformed. Respond with HTTP 400 Bad Request + the
    /// included message in `{"error": ...}`.
    BadRequest(String),
    /// Envelope is well-formed but no handler is registered for `job_code`.
    /// Respond with HTTP 404 Not Found + the included message in
    /// `{"error": ...}`.
    NotFound(String),
}

/// Options for [`ScheduledJobRunner::builder`].
pub struct RunnerOptions {
    pub lock_provider: Arc<dyn LockProvider>,
    pub lock_key: Box<dyn Fn(&ScheduledJobEnvelope) -> String + Send + Sync + 'static>,
    pub enforce_lock: bool,
    pub lock_ttl: Duration,
    pub on_error: Option<OnErrorHook>,
}

impl RunnerOptions {
    pub fn default_with_lock(lock_provider: Arc<dyn LockProvider>) -> Self {
        Self {
            lock_provider,
            lock_key: Box::new(|e| format!("scheduled-job:{}", e.job_code)),
            enforce_lock: true,
            lock_ttl: DEFAULT_LOCK_TTL,
            on_error: None,
        }
    }
}

/// Registry-of-handlers + envelope dispatcher.
///
/// Build via [`ScheduledJobRunner::builder`]; register handlers via
/// [`ScheduledJobRunner::handler`]; then expose the runner from your HTTP
/// route so each platform → SDK firing calls [`ScheduledJobRunner::process`].
pub struct ScheduledJobRunner {
    handlers: HashMap<String, BoxedHandler>,
    client: FlowCatalystClient,
    options: RunnerOptions,
}

impl ScheduledJobRunner {
    /// Start a builder. Pass the API client used for log/complete callbacks
    /// and a [`LockProvider`] (use `Arc::new(NoOpLockProvider)` if you don't
    /// need concurrency control).
    pub fn builder(
        client: FlowCatalystClient,
        lock_provider: Arc<dyn LockProvider>,
    ) -> ScheduledJobRunnerBuilder {
        ScheduledJobRunnerBuilder {
            client,
            handlers: HashMap::new(),
            options: RunnerOptions::default_with_lock(lock_provider),
        }
    }

    /// Process an inbound platform → SDK firing. Validates the envelope,
    /// kicks off the handler in the background, and returns a [`RunResult`]
    /// immediately. The actual handler execution + completion callback
    /// continues asynchronously via `tokio::spawn`.
    pub fn process(&self, envelope: serde_json::Value) -> RunResult {
        let env = match serde_json::from_value::<ScheduledJobEnvelope>(envelope) {
            Ok(e) => e,
            Err(err) => return RunResult::BadRequest(err.to_string()),
        };
        if !matches!(env.trigger_kind, TriggerKind::Cron | TriggerKind::Manual) {
            return RunResult::BadRequest("Invalid triggerKind".into());
        }
        let Some(handler) = self.handlers.get(&env.job_code).cloned() else {
            return RunResult::NotFound(format!(
                "No handler registered for code '{}'",
                env.job_code
            ));
        };

        let client = self.client.clone();
        let lock_provider = Arc::clone(&self.options.lock_provider);
        let enforce_lock = self.options.enforce_lock;
        let lock_ttl = self.options.lock_ttl;
        let lock_key = (self.options.lock_key)(&env);
        let on_error = self.options.on_error.clone();

        tokio::spawn(async move {
            run_in_background(
                env,
                handler,
                client,
                lock_provider,
                enforce_lock,
                lock_ttl,
                lock_key,
                on_error,
            )
            .await;
        });

        RunResult::Accepted
    }

    /// Listed registered codes (diagnostics only).
    pub fn registered_codes(&self) -> Vec<&str> {
        self.handlers.keys().map(String::as_str).collect()
    }
}

pub struct ScheduledJobRunnerBuilder {
    client: FlowCatalystClient,
    handlers: HashMap<String, BoxedHandler>,
    options: RunnerOptions,
}

impl ScheduledJobRunnerBuilder {
    /// Register a handler keyed by the job's `code`.
    ///
    /// Example:
    ///
    /// ```ignore
    /// .handler("daily-rollup", |ctx| Box::pin(async move {
    ///     ctx.log("starting", LogOptions::default()).await;
    ///     // … work …
    ///     Ok(serde_json::json!({ "processed": 42 }))
    /// }))
    /// ```
    pub fn handler<F>(mut self, code: impl Into<String>, handler: F) -> Self
    where
        F: Fn(HandlerContext) -> HandlerFuture + Send + Sync + 'static,
    {
        self.handlers.insert(code.into(), Arc::new(handler));
        self
    }

    pub fn lock_key<F>(mut self, f: F) -> Self
    where
        F: Fn(&ScheduledJobEnvelope) -> String + Send + Sync + 'static,
    {
        self.options.lock_key = Box::new(f);
        self
    }

    pub fn enforce_lock(mut self, enforce: bool) -> Self {
        self.options.enforce_lock = enforce;
        self
    }

    pub fn lock_ttl(mut self, ttl: Duration) -> Self {
        self.options.lock_ttl = ttl;
        self
    }

    pub fn on_error<F>(mut self, hook: F) -> Self
    where
        F: Fn(RunnerError, &ScheduledJobEnvelope) + Send + Sync + 'static,
    {
        self.options.on_error = Some(Arc::new(hook));
        self
    }

    pub fn build(self) -> ScheduledJobRunner {
        ScheduledJobRunner {
            handlers: self.handlers,
            client: self.client,
            options: self.options,
        }
    }
}

// ── Background dispatch ──────────────────────────────────────────────────

#[allow(clippy::too_many_arguments)]
async fn run_in_background(
    envelope: ScheduledJobEnvelope,
    handler: BoxedHandler,
    client: FlowCatalystClient,
    lock_provider: Arc<dyn LockProvider>,
    enforce_lock: bool,
    lock_ttl: Duration,
    lock_key: String,
    on_error: Option<OnErrorHook>,
) {
    let mut lock_handle: Option<crate::lock::LockHandle> = None;

    if enforce_lock {
        match lock_provider.acquire(&lock_key, lock_ttl).await {
            Ok(Some(h)) => lock_handle = Some(h),
            Ok(None) => {
                // Lock contention — skip this firing. If the job tracks
                // completion, mark as FAILURE with a clear reason; otherwise
                // just no-op (the instance will sit DELIVERED forever).
                if envelope.tracks_completion {
                    let req = InstanceCompleteRequest {
                        status: CompletionStatus::Failure,
                        result: Some(serde_json::json!({
                            "skipped": true,
                            "reason": "lock-held",
                        })),
                    };
                    if let Err(e) = client
                        .scheduled_jobs()
                        .complete_instance(&envelope.instance_id, &req)
                        .await
                    {
                        invoke_on_error(&on_error, RunnerError::CallbackFailed(e), &envelope);
                    }
                }
                return;
            }
            Err(e) => {
                invoke_on_error(
                    &on_error,
                    RunnerError::LockReleaseFailed(format!("acquire failed: {e}")),
                    &envelope,
                );
                return;
            }
        }
    }

    let ctx = HandlerContext {
        envelope: envelope.clone(),
        inner: Arc::new(HandlerContextInner {
            client: client.clone(),
            instance_id: envelope.instance_id.clone(),
            job_code: envelope.job_code.clone(),
            on_error: on_error.clone(),
        }),
    };

    let outcome = handler(ctx).await;

    if envelope.tracks_completion {
        let req = match &outcome {
            Ok(value) => InstanceCompleteRequest {
                status: CompletionStatus::Success,
                result: Some(sanitise_result(value.clone())),
            },
            Err(err) => InstanceCompleteRequest {
                status: CompletionStatus::Failure,
                result: Some(serde_json::json!({ "error": err.message })),
            },
        };
        if let Err(e) = client
            .scheduled_jobs()
            .complete_instance(&envelope.instance_id, &req)
            .await
        {
            invoke_on_error(&on_error, RunnerError::CallbackFailed(e), &envelope);
        }
    }

    if let Err(err) = &outcome {
        invoke_on_error(
            &on_error,
            RunnerError::HandlerPanicked(err.message.clone()),
            &envelope,
        );
    }

    if let Some(handle) = lock_handle {
        handle.release().await;
    }
}

fn invoke_on_error(
    hook: &Option<OnErrorHook>,
    err: RunnerError,
    envelope: &ScheduledJobEnvelope,
) {
    if let Some(h) = hook {
        h(err, envelope);
    }
}

fn sanitise_result(v: serde_json::Value) -> serde_json::Value {
    match serde_json::to_string(&v) {
        Ok(json) if json.len() > MAX_RESULT_BYTES => serde_json::json!({
            "truncated": true,
            "preview": json.chars().take(MAX_RESULT_BYTES).collect::<String>(),
        }),
        _ => v,
    }
}
